dpnp/dpnp_iface_mathematical.py

```diff
@@ -2742,6 +2778,181 @@ def gradient(f, *varargs, axis=None, edge_order=1):
+
+    fp : {dpnp.ndarray, usm_ndarray}
+        Input 1-D array. The y-coordinates of the data points,
+        same length as `xp`.
+
+    left : {None, scalar, dpnp.ndarray, usm_ndarray}, optional
+        Value to return for `x < xp[0]`.
+
+        Default: ``fp[0]``.
+
+    right : {None, scalar, dpnp.ndarray, usm_ndarray}, optional
+        Value to return for `x > xp[-1]`.
+
+        Default: ``fp[-1]``.
+
+    period : {None, scalar}, optional
+        A period for the x-coordinates. This parameter allows the proper
+        interpolation of angular x-coordinates. Parameters `left` and `right`
+        are ignored if `period` is specified.
+
+        Default: ``None``.
+
+    Returns
+    -------
+    y : {dpnp.ndarray, usm_ndarray}
+        The interpolated values, same shape as `x`.
+
+
+    Warnings
+    --------
+    The x-coordinate sequence is expected to be increasing, but this is not
+    explicitly enforced.  However, if the sequence `xp` is non-increasing,
+    interpolation results are meaningless.
+
+    Note that, since NaN is unsortable, `xp` also cannot contain NaNs.
+
+    A simple check for `xp` being strictly increasing is::
+
+        import dpnp as np
+        np.all(np.diff(xp) > 0)
+
+    Examples
+    --------
+    >>> import dpnp as np
+    >>> xp = np.array([1, 2, 3])
+    >>> fp = np.array([3 ,2 ,0])
+    >>> x = np.array([2.5])
+    >>> np.interp(x, xp, fp)
+    array([1.])
+    >>> x = np.array([0, 1, 1.5, 2.72, 3.14])
+    >>> np.interp(x, xp, fp)
+    array([3.  , 3.  , 2.5 , 0.56, 0.  ])
+    >>> x = np.array([3.14])
+    >>> UNDEF = -99.0
+    >>> np.interp(x, xp, fp, right=UNDEF)
+    array([-99.])
+
+    Interpolation with periodic x-coordinates:
+
+    >>> x = np.array([-180, -170, -185, 185, -10, -5, 0, 365])
+    >>> xp = np.array([190, -190, 350, -350])
+    >>> fp = np.array([5, 10, 3, 4])
+    >>> np.interp(x, xp, fp, period=360)
+    array([7.5 , 5.  , 8.75, 6.25, 3.  , 3.25, 3.5 , 3.75])
+
+    Complex interpolation:
+
+    >>> x = np.array([1.5, 4.0])
+    >>> xp = np.array([2,3,5])
+    >>> fp = np.array([1.0j, 0, 2+3j])
+    >>> np.interp(x, xp, fp)
+    array([0.+1.j , 1.+1.5j])
+
+    """
+
+    dpnp.check_supported_arrays_type(x, xp, fp)
+
+    if xp.ndim != 1 or fp.ndim != 1:
+        raise ValueError("xp and fp must be 1D arrays")
+    if xp.size != fp.size:
+        raise ValueError("fp and xp are not of the same length")
+    if xp.size == 0:
+        raise ValueError("array of sample points is empty")
+
+    usm_type, exec_q = get_usm_allocations([x, xp, fp])
+
+    x_dtype = dpnp.common_type(x, xp)
+    x_float_type = dpnp.default_float_type(exec_q)
+
+    if not dpnp.can_cast(x_dtype, x_float_type):
+        raise TypeError(
+            "Cannot cast array data from"
+            f" {x_dtype} to {x_float_type} "
+            "according to the rule 'safe'"
+        )
+
+    x = dpnp.asarray(x, dtype=x_float_type, order="C")
+    xp = dpnp.asarray(xp, dtype=x_float_type, order="C")
```

Comment:
dtype check is implemented as part of `astype`, so:
```suggestion
    x_float_type = dpnp.default_float_type(exec_q)
    x = dpnp.astype(x, x_float_type, order="C", casting="safe", copy=False)
    xp = dpnp.astype(xp, x_float_type, order="C", casting="safe", copy=False)
```